File: antia_numerical_methods/f77/sub8/lagurw.f
!     To calculate weights and abscissas of a Gauss-Laguerre quadrature formula
!
!     N : (input) Number of points in the required quadrature formula
!     ALP : (input) Exponent of x in weight function, w(x)=x**ALP*EXP(-x)
!     W : (output) Array of length N, which will contain the weights
!     A : (output) Array of length N containing the abscissas
!     IER : (output) Error parameter, IER=0 implies successful execution
!		IER=302 implies N.LE.0
!		IER=313 implies ALP.LE.-1
!		IER=321 implies that some coefficient becomes imaginary
!			during calculations.
!		In both these cases calculations are abandoned.
!		Other values may be set by TQL2
!     WK : Real array of length N*(N+2)+3*(N+1) used as scratch space
!
!     Required routines : GAUSRC, TQL2, GAMMA
 
      SUBROUTINE LAGURW(N,ALP,W,A,IER,WK)
      IMPLICIT REAL*8(A-H,O,P,R-Z)
      DIMENSION A(N),W(N),WK(3,*)
 
      IF(ALP.LE.-1.0) THEN
        IER=313
        RETURN
      ENDIF

      RI0=GAMMA(ALP+1.0)
      DO 1500 I=1,N+1
!     Coefficients of recurrence relation
        WK(1,I)=-(ALP+I)/I
        WK(2,I)=(ALP+2*I-1)*(ALP+I)/I
        WK(3,I)=(I-1.D0+ALP)**2*(ALP+I)/I
1500  CONTINUE
 
      CALL GAUSRC(N,W,A,WK,RI0,IER,WK(1,N+2))
      END

File: antia_numerical_methods@Beliavsky/c/fun/lagurw.c
/*     To calculate weights and abscissas of a Gauss-Laguerre quadrature formula

	N : (input) Number of points in the required quadrature formula
	ALP : (input) Exponent of x in weight function, w(x)=x**ALP*EXP(-x)
	W : (output) Array of length N, which will contain the weights
	A : (output) Array of length N containing the abscissas
		
	Error status is returned by the value of the function LAGURW.
		0 value implies successful execution
		302 implies N<= 0
		313 implies ALP<= -1
		321 implies that some coefficient becomes imaginary
			during calculations.
		In both these cases calculations are abandoned.
		Other values may be set by TQL2

	Required functions : GAUSRC, TQL2, GAMMA
*/

#include <math.h>
#include <stdlib.h>

int gausrc(int n, double w[], double a[], double *cof, double ri0);

int lagurw(int n, double alp, double w[], double a[])

{
	int i,ier;
	double ri0;
	double *wk;

	if(alp<= -1) return 313;

	ri0=gamma(alp+1.0);
	wk=(double *) calloc((size_t) (3*(n+1)), sizeof(double));

/*     Coefficients of recurrence relation */
	for(i=0; i<=n; ++i) {
		wk[i*3]=-(alp+i+1.0)/(i+1.0);
		wk[1+i*3]=(2*i+1.0+alp)*(alp+i+1.0)/(i+1.0);
		wk[2+i*3]=(i+alp)*(i+alp)*(i+1+alp)/(i+1.0);
	}
	
	ier=gausrc(n,w,a,wk,ri0);
	free(wk);
	return ier;
}
